### PR TITLE
Change torch.library.impl to accept a device string

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111661
* #111660
* #111310
* __->__ #111659
* #111380

torch.library.impl now accepts a device string (e.g. "cpu", "cuda"). It
still accepts DispatchKey strings, but we no longer document this, because
using arbitrary DispatchKeys is more for the power users.

We map the device string to a DispatchKey and then register the impl for
said DispatchKey. A user may also specify multiple device strings at once
or specify "types=default" to get a CompositeExplicitAutograd registration.

Test Plan:
- new tests

### CODE REVIEW

File: torch/library.py [hunk 1]
```diff
@@ -351,7 +379,7 @@ def impl_abstract(qualname, func=None, *, lib=None, _stacklevel=1):
         >>>     result = x.new_empty(shape, dtype=torch.int64)
         >>>     return result
         >>>
-        >>> @torch.library.impl(lib, "custom_nonzero", "CPU")
+        >>> @torch.library.impl("mylib::custom_nonzero", "cpu")
```
Comment (soulitzer) [inline torch/library.py:382]: With the new API, would it be best practice to omit passing lib all the time for simplicity, and rely on impl/define functions to create fragments behind the scenes? If that is the case should we accept the lib parameter at all, or are there other reasons for having it around?

Comment (soulitzer) [reply torch/library.py:382]: Could another way be just to have some kind of global mapping somewhere that remembers how to map ns to lib?

Comment (zou3519) [reply torch/library.py:382]: > With the new API, would it be best practice to omit passing lib all the time for simplicity, and rely on impl/define functions to create fragments behind the scenes?

Yup

>  If that is the case should we accept the lib parameter at all, or are there other reasons for having it around?

This is useful for testing. I don't expect anyone to actually use the lib parameter, but I'm keeping it around as an option.

Comment (soulitzer) [reply torch/library.py:382]: I see, thanks.

File: torch/csrc/utils/python_dispatch.cpp [hunk 2]
```diff
@@ -772,6 +782,16 @@ void initDispatchBindings(PyObject* module) {
     return at::functionalization::impl::commit_update(a);
   });
 
+  m.def("_dispatch_key_for_device", [](const std::string& device_type) {
+    auto device = c10::Device(device_type);
+    TORCH_CHECK(
+        !device.has_index(),
+        "Expected device_type string to not a device index; got ",
```
Comment (soulitzer) [inline torch/csrc/utils/python_dispatch.cpp:789]:
```suggestion
        "Expected device_type string to not have a device index; got ",
```

**Review by soulitzer — APPROVED** [review]

File: torch/csrc/utils/python_dispatch.cpp (hunk 2 above)
Comment (soulitzer) [reply torch/csrc/utils/python_dispatch.cpp:789]: also device_type -> device type? 
since we didn't actually name the argument to impl `device_type`

**Pull request closed without merging**